doc/build-overview.md

```diff
@@ -0,0 +1,251 @@
+  does not exist in any of the three parsed maps from the "wanted
+  compiler" step above, or that the package is present as an immutable
+  dependency from Hackage. If so, create an immutable dependency entry
+  with default flags, GHC options, and hidden status, and add this
+  package to the set of immutable package dependencies.
+* For all package names, ensure the package is in one of the four maps
+  we have, and if so add to either the dependency or project package
+  target set.
+* For all package name + component, ensure that the package is a
+  project package, and add that package + component to the set of
+  project targets.
+* Ensure that no target has been specified multiple times.
+
+We now have an update four package maps, a new set of dependency
+targets, and a new set of project package targets (potentially with
+specific components).
+
+## Apply named CLI flags
+
+Named CLI flags are applied to specific packages by updating the
+config in one of the four maps. If a flag is specified and no package
+is found, it's an error.
+
+## Apply CLI GHC options
+
+Apply GHC options from the command line to all _project package
+targets_. *FIXME* confirm that this is in fact the correct behavior.
+
+## Apply general flags (CLI and config value)
+
+*FIXME* figure out and document exactly which packages these will
+apply to.
+
+## Apply general GHC options
+
+*FIXME* list out the various choices here and which packages they
+apply to.
+
+## Determine snapshot hash
+
+Use some deterministic binary serialization and SHA256 thereof to get
+a hash of the following information:
+
+* Actual compiler (GHC version, path, *FIXME* probably some other
+  unique info from GHC, I've heard that `ghc --info` gives you
+  something)
+* Global database map
+* Immutable dependency map
+
+Motivation: Any package built from the immutable dependency map and
+installed in this database will never need to be rebuilt.
+
+*FIXME* Caveat: do we need to take profiling settings into account
+here? How about Haddock status?
+
+## Determine actual target components
+
+* Dependencies: "default" components (all libraries and executables)
+* Project packages:
+    * If specific components named: only those, plus any libraries present
+    * If no specific components, include the following:
+        * All libraries, always
+        * All executables, always
+        * All test suites, _if_ `--test` specified on command line
+        * All benchmarks, _if_ `--bench` specified on command line
+
+## Construct build plan
+
+* Applied to every target (project package or dependency)
+* Apply flags, platform, and actual GHC version to resolve
+  dependencies in any package analyzed
+* Include all library dependencies for all enabled components
+* Include all build tool dependencies for all enabled components
+  (using the fun backwards compat logic for `build-tools`)
+* Apply the logic recursively to come up with a full build plan
```

Comment:
A note here: one awkward thing about formulating build plan construction this way is that you might reach package p via an edge that says "I only need the library", and then later in the recursive traversal you reach p again where it says, "Oh, I also need the executable." So if you want to implement it this way, you have to have a notion where you can take a package that is already in your plan, and increase the components that are enabled for it. If you can truly assume that all packages support per-component builds, then you can just memoize over components and there isn't any problem; this might help.

Reply:
Is there some other approach to build plan construction which doesn't present this issue? I'm also not quite following your last sentence, can you clarify?

Reply:
> Is there some other approach to build plan construction which doesn't present this issue?

Well, there are a few alternate approaches, but they have their own tradeoffs.

1. You could say, recursive traversal operates over *components*, not packages. So you'll "visit" packages multiple times, but you'll only visit any given component once. The downside is that you'll end up having split up a package into multiple components to build, but some packages can't be built on a per-component basis (e.g., a package with a Custom setup) so you'll have to agglomerate them all back together before you actually do a build. (This is what I was alluding to in the last sentence.)

2. You could process per-package, but say, "Actually, I'm going to assume that *all* components are requested." So you will never have to improve because you always greedily configured every component, so it must be available (or be impossible). The downside is that you spent a whole bunch of work figuring out a build plan to build everything, which is not actually what you want, so you're going to have to prune it down after the fact. You probably don't want to be recomputing this every time, so you had better cache the plan. (By the way, this is what cabal new-build does).

Reply:
Makes sense. I think I'll leave the doc as-is for now, and update it during implementation. I have a feeling we'll end up closer to your (1).